File: Python_백준/2178_미로탐색.py
#2178

import queue
N, M = map(int, input.split())

Map = [[False for x in range(M)] for y in range(M)] #장애물 유무

for y in range(N):
    r = input()
    for x in range(M):
        if r[x] == 0:
            Map[y][x] = True

q = queue.Queue()

q.put([0, 0, 0])#좌표, cnt
dir = [[-1, 0], [1, 0], [0, -1], [0, 1]]#상하좌우
visit = [[False for x in range(M)] for y in range(M)]
while not q.empty():
    t = q.get()

    if t[:2] == [M -1, N - 1]:
        break

    for d in dir:
        x = t[0] + d[0]
        y = t[1] + d[1]
        visit[y][x] = True
        q.put([x, y, t[2] + 1])

File: Cpp_백준/2178_미로탐색.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>
using namespace std;

struct Dir{
    int x;
    int y;
};

Dir D[4] = {{0, -1}, {1, 0}, {0, 1}, {-1, 0}};
vector<int> *graph, *visit;
queue<Dir> q;
int n , m; //n : height , m : width

void bfs(int nodex, int nodey){
    Dir t;

    q.push({nodex, nodey});
    visit[nodey].at(nodex) = 1;
    while(!q.empty()){
        t = q.front();
        for (int i = 0; i < 4; i++){
            if (t.x + D[i].x >= 0 && t.x + D[i].x < m && t.y + D[i].y >= 0 && t.y + D[i].y < n){
                if (visit[t.y + D[i].y].at(t.x + D[i].x) == -1 && graph[t.y + D[i].y].at(t.x + D[i].x) == 1){
                    q.push({t.x + D[i].x, t.y + D[i].y});
                    visit[t.y + D[i].y].at(t.x + D[i].x) = visit[t.y].at(t.x) + 1;
                }
            }
        }
        q.pop();
    }
}

int main(){
    string tmp;

    cin >> n >> m;
    graph = new vector<int>[n];
    visit = new vector<int>[n];

    for (int i = 0; i < n; i++){
        cin >> tmp;
        for (int j = 0; j < m; j++){
            visit[i].push_back(-1);
            graph[i].push_back(tmp[j] - '0');
        }
    }

    bfs(0, 0);
    cout << visit[n - 1].at(m - 1) << endl;
}